x/leverage/keeper/keeper.go

```diff
@@ -271,3 +271,166 @@ func (k Keeper) GetCollateralSetting(ctx sdk.Context, borrowerAddr sdk.AccAddres
+) error {
+	if !repayment.IsValid() {
+		return sdkerrors.Wrap(types.ErrInvalidAsset, repayment.String())
+	}
+	if !k.IsAcceptedUToken(ctx, rewardDenom) {
+		return sdkerrors.Wrap(types.ErrInvalidAsset, rewardDenom)
+	}
+
+	// Get total borrowed by borrower (all denoms)
+	borrowed, err := k.GetBorrowerBorrows(ctx, borrowerAddr)
+	if err != nil {
+		return err
+	}
+
+	// Get borrower uToken balances, for all uToken denoms enabled as collateral
+	collateral := k.GetBorrowerCollateral(ctx, borrowerAddr)
+	if err != nil {
+		return err
+	}
+
+	// Use oracle helper functions to find total borrowed value in USD
+	borrowValue, err := k.TotalPrice(ctx, borrowed)
+	if err != nil {
+		return err
+	}
+
+	// Use collateral weights to compute borrow limit from enabled collateral
+	borrowLimit, err := k.CalculateBorrowLimit(ctx, collateral)
+	if err != nil {
+		return err
+	}
+
+	// Confirm borrower's eligibility for liquidation
+	if borrowLimit.GTE(borrowValue) {
+		return sdkerrors.Wrap(types.ErrLiquidationIneligible, borrowerAddr.String())
+	}
+
+	// Get reward-specific incentive and dynamic close factor
+	baseRewardDenom := k.FromUTokenToTokenDenom(ctx, rewardDenom)
+	liquidationIncentive, closeFactor, err := k.LiquidationParams(ctx, baseRewardDenom, borrowValue, borrowLimit)
+	if err != nil {
+		return err
+	}
+
+	// Repayment cannot exceed liquidator's available balance
+	liquidatorBalance := k.bankKeeper.GetBalance(ctx, liquidatorAddr, repayment.Denom)
+	if repayment.Amount.GTE(liquidatorBalance.Amount) {
+		repayment.Amount = liquidatorBalance.Amount
+	}
+
+	// Repayment cannot exceed borrower's borrowed amount of selected denom
+	if repayment.Amount.GTE(borrowed.AmountOf(repayment.Denom)) {
+		repayment.Amount = borrowed.AmountOf(repayment.Denom)
+	}
+
+	// Repayment cannot exceed borrowed value * close factor
+	repayValue, err := k.Price(ctx, repayment)
+	if err != nil {
+		return err
+	}
+	if repayValue.GTE(borrowValue.Mul(closeFactor)) {
+		maxRepayValue := borrowValue.Mul(closeFactor)
+		// repayment *= (maxRepayValue / repayValue)
+		repayment.Amount = repayment.Amount.ToDec().Mul(maxRepayValue).Quo(repayValue).TruncateInt()
+		repayValue = maxRepayValue
+	}
+
+	// Given repay denom and amount, use oracle to find equivalent amount of rewardDenom's base asset
+	baseReward, err := k.EquivalentValue(ctx, repayment, baseRewardDenom)
+	if err != nil {
+		return err
+	}
+
+	// Convert reward tokens back to uTokens
+	reward, err := k.ExchangeTokens(ctx, baseReward)
+	if err != nil {
+		return err
+	}
+
+	// Apply liquidation incentive
+	reward.Amount = reward.Amount.ToDec().Mul(sdk.OneDec().Add(liquidationIncentive)).TruncateInt()
+
+	// Reward amount cannot exceed available collateral
+	if reward.Amount.GTE(collateral.AmountOf(rewardDenom)) {
+		// only pay what can be correctly compensated
+		repayment.Amount = repayment.Amount.Mul(collateral.AmountOf(rewardDenom)).Quo(reward.Amount)
```

Comment:
I find that modifying `repayment.Amount` in two different places to be confusing. Can we comment on why we do this and how it differs from above?

Reply:
Comment updated, LMK if it's good this time around